validation/src/verifier/key_verifier.rs

```diff
@@ -0,0 +1,76 @@
+use anyhow::{anyhow, bail, Context, Result};
+use base64::engine::{self, Engine as _};
+use ceramic_core::Jwk;
+use ceramic_event::unvalidated::signed::Envelope;
+use tracing::{debug, warn};
+
+#[async_trait::async_trait]
+pub trait Verifier {
+    /// Verify the signature. Optoinally requiring it matches a delegated KID
+    async fn verify_signature(&self, delegated_kid: Option<&str>) -> Result<()>;
+}
+
+#[async_trait::async_trait]
+impl Verifier for Envelope {
+    async fn verify_signature(&self, delegated_kid: Option<&str>) -> Result<()> {
+        let signature = self
+            .signature()
+            .first()
+            .ok_or_else(|| anyhow!("missing signature on signed event"))?;
+        let jws_header = self.jws_header().context("event envelope is not a jws")?;
+        let protected = signature
+            .protected
+            .as_ref()
+            .ok_or_else(|| anyhow!("missing protected field"))?;
+
+        let did = &jws_header
+            .key_id
+            .ok_or_else(|| anyhow!("missing jws kid"))?;
+
+        verify_did_jws(
+            did,
+            protected.as_slice(),
+            self.payload().as_slice(),
+            jws_header.algorithm,
+            signature.signature.as_slice(),
+            delegated_kid,
+        )
+        .await
+    }
+}
+
+pub(crate) async fn verify_did_jws(
+    did: &str,
+    header: &[u8],
+    payload: &[u8],
+    alg: ssi::jwk::Algorithm,
+    signature: &[u8],
+    delegated_kid: Option<&str>,
```

Comment:
why is this called delegated_kid, rather than just a kid?

Reply:
this is related to step 4 of [cacao verification](https://developers.ceramic.network/docs/protocol/js-ceramic/accounts/object-capabilities#verification).. even if the cacao was signed and the envelope was signed correctly, the cacao had to actually delegate to the key that signed the envelope. I don't care for this API signature but it's mostly private/implementation details right now.